core/src/main/scala/org/apache/spark/BarrierTaskContext.scala

```diff
@@ -163,6 +131,71 @@ class BarrierTaskContext private[spark] (
       timerTask.cancel()
       timer.purge()
     }
+    messagesSerialized
+  }
+
+  /**
+   * :: Experimental ::
+   * Sets a global barrier and waits until all tasks in this stage hit this barrier. Similar to
+   * MPI_Barrier function in MPI, the barrier() function call blocks until all tasks in the same
+   * stage have reached this routine.
+   *
+   * CAUTION! In a barrier stage, each task must have the same number of barrier() calls, in all
+   * possible code branches. Otherwise, you may get the job hanging or a SparkException after
+   * timeout. Some examples of '''misuses''' are listed below:
+   * 1. Only call barrier() function on a subset of all the tasks in the same barrier stage, it
+   * shall lead to timeout of the function call.
+   * {{{
+   *   rdd.barrier().mapPartitions { iter =>
+   *       val context = BarrierTaskContext.get()
+   *       if (context.partitionId() == 0) {
+   *           // Do nothing.
+   *       } else {
+   *           context.barrier()
+   *       }
+   *       iter
+   *   }
+   * }}}
+   *
+   * 2. Include barrier() function in a try-catch code block, this may lead to timeout of the
+   * second function call.
+   * {{{
+   *   rdd.barrier().mapPartitions { iter =>
+   *       val context = BarrierTaskContext.get()
+   *       try {
+   *           // Do something that might throw an Exception.
+   *           doSomething()
+   *           context.barrier()
+   *       } catch {
+   *           case e: Exception => logWarning("...", e)
+   *       }
+   *       context.barrier()
+   *       iter
+   *   }
+   * }}}
+   */
+  @Experimental
+  @Since("2.4.0")
+  def barrier(): Unit = {
+    runBarrier(RequestMethod.BARRIER)
+    ()
+  }
+
+  /**
+   * :: Experimental ::
+   * Blocks until all tasks in the same stage have reached this routine.  Each task passes in
+   * a message and returns with a list of all the messages passed in by each of those tasks.
+   *
+   * CAUTION! The allGather method requires the same precautions as the barrier method
+   */
+  @Experimental
+  @Since("3.0.0")
+  def allGather(message: Array[Byte]): ArrayBuffer[Array[Byte]] = {
```

Comment:
Is it better to directly use `String` as its input/return type?

 e.g. `def allGather(message: String): ArrayBuffer[String]`, which should be more friendly to end user.

Reply:
This was the previous implementation but changed to the `Array[Byte]` API so that users who are serializing non-strings don't have decreased performance. It's definitely a tradeoff because as you say, it's far less convenient for many users. @jiangxb1987 what do you think?

Reply:
Talked to @mengxr  offline, the targeted use cases are not performance sensitive, we don't expect to pass long strings through the `allGather` method, thus I'm fine to use String as the input/return type.

Reply:
But please also add comment to explain the design decision :)